### PR TITLE
Move helper functions for scaling out BasePool to library

### PR DESCRIPTION
**Pull request by TomAFrench**
This seems like a sensible way to shrink BasePool prior to it being vendored.

```
    // Upscale rounding wouldn't necessarily always go in the same direction: in a swap for example the balance of
    // token in should be rounded up, and that of token out rounded down. This is the only place where we round in
    // the same direction for all amounts, as the impact of this rounding is expected to be minimal.
```
This comment in `_upscale` seems concerning however, should we split this into `_upscaleUp` and `_upscaleDown`?

### CODE REVIEW

File: pkg/solidity-utils/contracts/helpers/ScalingHelpers.sol [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+pragma solidity ^0.7.0;
+
+import "../math/FixedPoint.sol";
+import "./InputHelpers.sol";
+
+// solhint-disable
+
```
Comment (nventuro) [inline pkg/solidity-utils/contracts/helpers/ScalingHelpers.sol:21]: It'd be good to have a short blurb on what scaling means and why we do it

File: pkg/solidity-utils/contracts/helpers/ScalingHelpers.sol [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+pragma solidity ^0.7.0;
+
+import "../math/FixedPoint.sol";
+import "./InputHelpers.sol";
+
+// solhint-disable
```
[the remark this answers is not in the document]
Comment (TomAFrench) [reply pkg/solidity-utils/contracts/helpers/ScalingHelpers.sol:20]: solhint doesn't seem to like free functions as it wants us to provide a visibility (which we can't). It also wants not underscore but I wanted to keep the function names the same.
```
 26:1  error  '_upscale' should not start with _             private-vars-leading-underscore
 26:1  error  Explicitly mark visibility in function         func-visibility
```

File: pkg/solidity-utils/contracts/helpers/ScalingHelpers.sol [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+/**
+ * @dev Applies `scalingFactor` to `amount`, resulting in a larger or equal value depending on whether it needed
+ * scaling or not.
+ */
+function _upscale(uint256 amount, uint256 scalingFactor) pure returns (uint256) {
+    // Upscale rounding wouldn't necessarily always go in the same direction: in a swap for example the balance of
+    // token in should be rounded up, and that of token out rounded down. This is the only place where we round in
+    // the same direction for all amounts, as the impact of this rounding is expected to be minimal.
+    return FixedPoint.mulDown(amount, scalingFactor);
+}
```
[the remark this answers is not in the document]
Comment (TomAFrench) [reply pkg/solidity-utils/contracts/helpers/ScalingHelpers.sol:33]: Ok let's leave it as is and revisit later once ManagedPool is flat.

File: pkg/solidity-utils/contracts/helpers/ScalingHelpers.sol (hunk 2 above)
[the remark this answers is not in the document]
Comment (nventuro) [reply pkg/solidity-utils/contracts/helpers/ScalingHelpers.sol]: Off with its head.

File: pkg/solidity-utils/contracts/helpers/ScalingHelpers.sol (hunk 3 above)
[the remark this answers is not in the document]
Comment (nventuro) [reply pkg/solidity-utils/contracts/helpers/ScalingHelpers.sol:36]: Created https://github.com/balancer-labs/balancer-v2-monorepo/issues/1744

**Review by nventuro — APPROVED** [review]

**Pull request merged**